### PR TITLE
PCA9685 driver

### PR DESCRIPTION
**Pull request by nixward**
Also adds PCA9685 based Adafruit PWM/Servo shield

### CODE REVIEW

File: drivers/pwm/pwm_pca9685.c [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+{
+	const struct pca9685_config *config = dev->config;
+	uint8_t value;
+	int ret;
+
+	ret = i2c_write_read_dt(&config->i2c, &addr, sizeof(addr), &value, sizeof(value));
+	if (ret != 0) {
+		LOG_ERR("I2C write [0x%02X]=0x%02X: %d", addr, value, ret);
+	}
+	return value;
```
Comment (XenuIsWatching) [inline drivers/pwm/pwm_pca9685.c:71]: How would you know if an error occurred while reading it? It would be better if a pointer was passed for the value to be written to and the `ret` to be the return value

Comment (nixward) [reply drivers/pwm/pwm_pca9685.c:71]: Good point, I'll fix up set cycles and init functions as well.

**Comment by carlescufi** [comment]
@gmarull could you take a look?

File: drivers/pwm/Kconfig [hunk 2]
```diff
@@ -14,6 +14,12 @@ module = PWM
 module-str = pwm
 source "subsys/logging/Kconfig.template.log_config"
 
+config PWM_INIT_PRIORITY
+	int "PWM initialization priority"
+	default 70
+	help
+	  System initialization priority for PWM drivers.
```
Comment (gmarull) [inline drivers/pwm/Kconfig:21]: this needs to go into a separate commit (or PR) and all drivers should be updated accordingly.

File: drivers/pwm/Kconfig.pca9685 [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#
+# Copyright (c) 2022 Nick Ward <nix.ward@gmail.com>
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+config PWM_PCA9685
+	bool "PCA9685 16-channel, 12-bit PWM Fm+ I2C-bus LED controller"
+	depends on I2C
+	help
```
Comment (gmarull) [inline drivers/pwm/Kconfig.pca9685:10]: please use `DT_HAS_*` helpers, and `select I2C`

File: drivers/pwm/pwm_pca9685.c [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+static void set_pre_scale(const struct device *dev, uint8_t value)
+{
+	uint8_t mode1;
+
+	/* Unblock write to PRE_SCALE by setting SLEEP bit to logic 1 */
+	set_reg(dev, ADDR_MODE1, AUTO_INC | SLEEP);
+
+	mode1 = get_reg(dev, ADDR_MODE1);
+	if ((mode1 & RESTART) == 0x00) {
+		LOG_ERR("RESTART bit should be set");
```
Comment (gmarull) [inline drivers/pwm/pwm_pca9685.c:83]: error but continue execution?

File: drivers/pwm/pwm_pca9685.c [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+	}
+
+	set_reg(dev, ADDR_PRE_SCALE, value);
+
+	/* Clear SLEEP bit - See datasheet section 7.3.1.1 */
+	set_reg(dev, ADDR_MODE1, AUTO_INC);
+	k_sleep(K_USEC(500));
```
Comment (gmarull) [inline drivers/pwm/pwm_pca9685.c:90]: magic sleep, define/document?

File: drivers/pwm/pwm_pca9685.c [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+	/* Restart */
+	set_reg(dev, ADDR_MODE1, AUTO_INC | RESTART);
```
Comment (gmarull) [inline drivers/pwm/pwm_pca9685.c:93]: error handling is ignored throughout this function

File: drivers/pwm/pwm_pca9685.c [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+static int pca9685_set_cycles(const struct device *dev, uint32_t pwm, uint32_t period_count,
+			      uint32_t pulse_count, pwm_flags_t flags)
+{
+	const struct pca9685_config *config = dev->config;
+	struct pca9685_data *data = dev->data;
+	uint8_t buf[] = {0x00, 0x00, 0x00, 0x00, 0x00};
```
Comment (gmarull) [inline drivers/pwm/pwm_pca9685.c:101]: `[5] = { 0 }`

File: drivers/pwm/pwm_pca9685.c [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+	uint32_t led_off_count;
+	int32_t pre_scale;
+
+	if (pwm >= PCA9685_CHANNEL_CNT) {
```
Comment (gmarull) [inline drivers/pwm/pwm_pca9685.c:105]: please use `channel` naming, `pwm` is no longer used.

File: include/zephyr/drivers/pwm/pca9685.h [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+ * When using the internal oscillator this corresponds to an
+ * update rate of 24 Hz
+ */
+#define PCA9685_PWM_PERIOD_COUNT_MAX PCA9685_PWM_PERIOD_COUNT_PS(PCA9685_PRE_SCALE_MAX)
+
+/**
+ * When using the internal oscillator this corresponds to an
+ * update rate of 200 Hz
+ */
+#define PCA9685_PWM_PERIOD_COUNT_DEFAULT PCA9685_PWM_PERIOD_COUNT_PS(PCA9685_PRE_SCALE_DEFAULT)
```
Comment (gmarull) [inline include/zephyr/drivers/pwm/pca9685.h:49]: why is all that placed in a public header?

File: boards/shields/adafruit_16_ch_pwm_servo/Kconfig.defconfig [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+if SHIELD_ADAFRUIT_16_CH_PWM_SERVO
+
+config I2C
+	default y
+
+config PWM_PCA9685
+	default y
+
+endif # SHIELD_ADAFRUIT_16_CH_PWM_SERVO
```
Comment (gmarull) [inline boards/shields/adafruit_16_ch_pwm_servo/Kconfig.defconfig:12]: this is redundant, please check how drivers use DT_HAS* and select

File: boards/shields/adafruit_16_ch_pwm_servo/adafruit_16_ch_pwm_servo.overlay [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+		#pwm-cells = <3>;
+	};
+};
+
+/ {
+	pwmleds {
+		status = "okay";
+		compatible = "pwm-leds";
+
+		s_led0: s_led_0 {
```
Comment (gmarull) [inline boards/shields/adafruit_16_ch_pwm_servo/adafruit_16_ch_pwm_servo.overlay:23]: node names should use `-`, `s_led0: s-led-0`

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/pwm/pwm_pca9685.c (hunk 1 above)
Comment (nixward) [reply drivers/pwm/pwm_pca9685.c]: Fixed

File: drivers/pwm/Kconfig (hunk 2 above)
Comment (nixward) [reply drivers/pwm/Kconfig:21]: I've updated this in a separate commit

File: drivers/pwm/Kconfig.pca9685 (hunk 3 above)
Comment (nixward) [reply drivers/pwm/Kconfig.pca9685]: Updated

File: drivers/pwm/pwm_pca9685.c (hunk 4 above)
Comment (nixward) [reply drivers/pwm/pwm_pca9685.c:104]: Fixed, it now exits execution with an error code.

File: drivers/pwm/pwm_pca9685.c (hunk 5 above)
Comment (nixward) [reply drivers/pwm/pwm_pca9685.c]: Added define `OSCILLATOR_STABILIZE`

File: drivers/pwm/pwm_pca9685.c (hunk 6 above)
Comment (nixward) [reply drivers/pwm/pwm_pca9685.c]: Fixed

File: drivers/pwm/pwm_pca9685.c (hunk 8 above)
Comment (nixward) [reply drivers/pwm/pwm_pca9685.c]: Updated

File: include/zephyr/drivers/pwm/pca9685.h (hunk 9 above)
Comment (nixward) [reply include/zephyr/drivers/pwm/pca9685.h]: The intent of these macros was to provide to application code the PCA9685 PWM period default and limits of the device.

File: boards/shields/adafruit_16_ch_pwm_servo/Kconfig.defconfig (hunk 10 above)
Comment (nixward) [reply boards/shields/adafruit_16_ch_pwm_servo/Kconfig.defconfig]: Removed

File: drivers/pwm/Kconfig (hunk 2 above)
Comment (nixward) [reply drivers/pwm/Kconfig:21]: @gmarull I've moved it into it's own PR now.

File: include/zephyr/drivers/pwm/pca9685.h (hunk 9 above)
Comment (jfischer-no) [reply include/zephyr/drivers/pwm/pca9685.h]: > The intent of these macros was to provide to application code the PCA9685 PWM period default and limits of the device.

That should be handled by the driver or driver API, this header should not be public.

File: drivers/pwm/pwm_pca9685.c [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+	}
+
+out:
+	k_mutex_unlock(&data->mutex);
+
+	return ret;
+}
+
+static int pca9685_set_cycles(const struct device *dev, uint32_t channel, uint32_t period_count,
+			      uint32_t pulse_count, pwm_flags_t flags)
```
Comment (jfischer-no) [inline drivers/pwm/pwm_pca9685.c:133]: Can you please follow the "preferred limit on the length of a single line" coding style everywhere, and not just blindly apply clang-format?

```c
static int pca9685_set_cycles(const struct device *dev,
			      uint32_t channel, uint32_t period_count,
			      uint32_t pulse_count, pwm_flags_t flags)
```

File: include/zephyr/drivers/pwm/pca9685.h [hunk 13]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2022 Nick Ward <nix.ward@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_PWM_PCA9685_H_
```
Comment (jfischer-no) [inline include/zephyr/drivers/pwm/pca9685.h:7]: This header file should not be public.

File: boards/shields/adafruit_16_ch_pwm_servo/adafruit_16_ch_pwm_servo.overlay [hunk 14]
```diff
@@ -0,0 +1,72 @@
+/*
```
Comment (jfischer-no) [inline boards/shields/adafruit_16_ch_pwm_servo/adafruit_16_ch_pwm_servo.overlay:1]: I think shield directory should be just `boards/shields/pca9685`.

servo in `adafruit_16_ch_pwm_servo.overlay` is a bit confusing even it is the name, maybe it is better to call it `pca9658_generic_pwmled` shield?

File: boards/shields/adafruit_16_ch_pwm_servo/adafruit_16_ch_pwm_servo.overlay [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+ * Copyright (c) 2022 Nick Ward <nix.ward@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&arduino_i2c {
+	status = "okay";
+
+	pca9685: pca9685@40 {
```
Comment (jfischer-no) [inline boards/shields/adafruit_16_ch_pwm_servo/adafruit_16_ch_pwm_servo.overlay:10]: device nodelabel should be `<device>_<shield>`, see https://docs.zephyrproject.org/latest/hardware/porting/shields.html

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: dts/bindings/pwm/nxp,pca9685.yaml [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+      The 16 LEDn outputs are configured with an open-drain structure.
+      Otherwise they are configured with a totem pole structure.
+
+  och-on-ack:
+    type: boolean
+    description:
+      Outputs change on ACK.  Otherwise the outputs change on STOP
+      command.
+
+  invrt:
```
Comment (galak) [inline dts/bindings/pwm/nxp,pca9685.yaml:26]: Rename to `invert` to match linux binding

File: dts/bindings/pwm/nxp,pca9685.yaml [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+compatible: "nxp,pca9685"
+
+include: [pwm-controller.yaml, i2c-device.yaml, base.yaml]
+
+properties:
+  reg:
+    required: true
+
+  outdrv_open_drain:
```
Comment (galak) [inline dts/bindings/pwm/nxp,pca9685.yaml:14]: rename to `open-drain` to match linux binding.

File: dts/bindings/pwm/nxp,pca9685.yaml [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+      command.
+
+  invrt:
+    type: boolean
+    description:
+      Output logic state inverted. Value to use when no external driver
+      used.
+
+  "#pwm-cells":
+    const: 3
```
Comment (galak) [inline dts/bindings/pwm/nxp,pca9685.yaml:33]: change to 2

File: dts/bindings/pwm/nxp,pca9685.yaml [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+
+pwm-cells:
+  - channel
+  - period
+  - flags
```
Comment (galak) [inline dts/bindings/pwm/nxp,pca9685.yaml:38]: drop flags from here.

**Review by galak — DISMISSED** [review]
rename compatiable to `nxp,pca9685-pwm` to match:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/pwm/nxp,pca9685-pwm.txt


Rename properties to match those described in that file.

File: drivers/pwm/pwm_pca9685.c (hunk 12 above)
Comment (nixward) [reply drivers/pwm/pwm_pca9685.c]: Done

File: include/zephyr/drivers/pwm/pca9685.h (hunk 13 above)
Comment (nixward) [reply include/zephyr/drivers/pwm/pca9685.h]: I've removed it

File: boards/shields/adafruit_16_ch_pwm_servo/adafruit_16_ch_pwm_servo.overlay (hunk 14 above)
Comment (nixward) [reply boards/shields/adafruit_16_ch_pwm_servo/adafruit_16_ch_pwm_servo.overlay]: I've shortened the board name to `adafruit_pca9685`

File: boards/shields/adafruit_16_ch_pwm_servo/adafruit_16_ch_pwm_servo.overlay (hunk 15 above)
Comment (nixward) [reply boards/shields/adafruit_16_ch_pwm_servo/adafruit_16_ch_pwm_servo.overlay]: Done

File: dts/bindings/pwm/nxp,pca9685.yaml (hunk 16 above)
Comment (nixward) [reply dts/bindings/pwm/nxp,pca9685.yaml]: Done

File: dts/bindings/pwm/nxp,pca9685.yaml (hunk 17 above)
Comment (nixward) [reply dts/bindings/pwm/nxp,pca9685.yaml]: Done

File: dts/bindings/pwm/nxp,pca9685.yaml (hunk 18 above)
Comment (nixward) [reply dts/bindings/pwm/nxp,pca9685.yaml]: Done

File: dts/bindings/pwm/nxp,pca9685.yaml (hunk 19 above)
Comment (nixward) [reply dts/bindings/pwm/nxp,pca9685.yaml]: Done

**Comment by nixward** [comment]
@jfischer-no @galak @gmarull could you please have another look, thanks

File: drivers/pwm/pwm_pca9685.c (hunk 12 above)
Comment (jfischer-no) [reply drivers/pwm/pwm_pca9685.c]: Actually not, please apply "preferred limit on the length of a single line" coding style everywhere it makes sense.

Comment (nixward) [reply drivers/pwm/pwm_pca9685.c]: Apologies I missed that, can you please check again @jfischer-no

**Review by jfischer-no — DISMISSED** [review]
Tested on hardware, :+1:

**Comment by nixward** [comment]
@galak could you please check again?

File: drivers/pwm/pwm_pca9685.c [hunk 20]
```diff
@@ -0,0 +1,288 @@
...
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(pwm_pca9685, CONFIG_PWM_LOG_LEVEL);
+
+#define OSC_CLOCK_MAX	  50000000
+#define CHANNEL_CNT	  16
+
+#define ADDR_MODE1	  0x00
+#define SLEEP		  BIT(4)
```
Comment (gmarull) [inline drivers/pwm/pwm_pca9685.c:21]: missing include file defining `BIT`

File: drivers/pwm/pwm_pca9685.c [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+		buf[2] = LED_FULL_ON;
+	}
+
+	return i2c_write_dt(&config->i2c, buf, sizeof(buf));
+}
+
+static int pca9685_get_cycles_per_sec(const struct device *dev,
+				      uint32_t channel, uint64_t *cycles)
+{
+	ARG_UNUSED(channel);
```
Comment (henrikbrixandersen) [inline drivers/pwm/pwm_pca9685.c:224]:
```suggestion
	ARG_UNUSED(dev);
	ARG_UNUSED(channel);
```

File: drivers/pwm/pwm_pca9685.c [hunk 22]
```diff
@@ -0,0 +1,288 @@
...
+static int pca9685_init(const struct device *dev)
+{
+	const struct pca9685_config *config = dev->config;
+	struct pca9685_data *data = dev->data;
+	uint8_t value;
+	int ret;
+
+	(void)k_mutex_init(&data->mutex);
+
+	if (!device_is_ready(config->i2c.bus)) {
```
Comment (henrikbrixandersen) [inline drivers/pwm/pwm_pca9685.c:245]:
```suggestion
	if (!i2c_is_ready_dt(config->i2c)) {
```

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks good. Only a few minor nits:

File: drivers/pwm/pwm_pca9685.c (hunk 20 above)
Comment (nixward) [reply drivers/pwm/pwm_pca9685.c:22]: Thanks, included

File: drivers/pwm/pwm_pca9685.c (hunk 21 above)
Comment (nixward) [reply drivers/pwm/pwm_pca9685.c:226]: Thanks, done

File: drivers/pwm/pwm_pca9685.c (hunk 22 above)
Comment (nixward) [reply drivers/pwm/pwm_pca9685.c]: Done

**Review by gmarull — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]
